--- a/drivers/clk/bcm/clk-bcm63xx-gate.c
+++ b/drivers/clk/bcm/clk-bcm63xx-gate.c
@@ -184,6 +184,25 @@ static const struct clk_bcm63xx_table_en
 	{ },
 };
 
+static const struct clk_bcm63xx_table_entry bcm63268_timer_clocks[] = {
+	{ .name = "ephy1-timer", .bit = 0, },
+	{ .name = "ephy2-timer", .bit = 1, },
+	{ .name = "ephy3-timer", .bit = 2, },
+	{ .name = "gphy-timer", .bit = 3, },
+	{ .name = "dsl-timer", .bit = 4, },
+	{ .name = "wo_ephy-timer", .bit = 5, },
+	{ .name = "wo_dsl-timer", .bit = 6, },
+	{ .name = "fap1-timer", .bit = 11, },
+	{ .name = "fap2-timer", .bit = 15, },
+	{ .name = "uto_50-timer", .bit = 16, },
+	{ .name = "uto_ext-timer", .bit = 17, },
+	{ .name = "usb_ref-timer", .bit = 18, },
+	{ .name = "sw_rst-timer", .bit = 29, },
+	{ .name = "hw_rst-timer", .bit = 30, },
+	{ .name = "port_rst-timer", .bit = 31, },
+	{ },
+};
+
 static int clk_bcm63xx_probe(struct platform_device *pdev)
 {
 	const struct clk_bcm63xx_table_entry *entry, *table;
@@ -267,6 +286,7 @@ static const struct of_device_id clk_bcm
 	{ .compatible = "brcm,bcm6362-clocks", .data = &bcm6362_clocks, },
 	{ .compatible = "brcm,bcm6368-clocks", .data = &bcm6368_clocks, },
 	{ .compatible = "brcm,bcm63268-clocks", .data = &bcm63268_clocks, },
+	{ .compatible = "brcm,bcm63268-timer-clocks", .data = &bcm63268_timer_clocks, },
 	{ }
 };
 
